crates/bevy_animation/src/lib.rs

```diff
@@ -479,173 +539,99 @@ impl AnimationPlayer {
-            update_transitions(&mut player, &time);
-            run_animation_player(
-                root,
-                player,
-                &time,
-                &animations,
-                &names,
-                &transforms,
-                &morphs,
-                maybe_parent,
-                &parents,
-                &children,
-            );
+        .for_each(|(id, target, name, (transform, morph_weights))| {
+            let mut target_context = AnimationTargetContext {
+                id,
+                target,
+                name,
+                transform,
+                morph_weights,
+            };
+
+            let Ok(player) = players.get(target.player) else {
+                error!(
+                    "Couldn't find the animation player {:?} for {:?} ({:?})",
+                    target.player, target_context.id, target_context.name,
+                );
+                return;
+            };
+
+            player.animation.apply(&clips, 1.0, &mut target_context);
+
+            for transition in &player.transitions {
+                transition
+                    .animation
+                    .apply(&clips, transition.current_weight, &mut target_context);
+            }
         });
 }
 
-#[allow(clippy::too_many_arguments)]
-fn run_animation_player(
-    root: Entity,
-    mut player: Mut<AnimationPlayer>,
-    time: &Time,
-    animations: &Assets<AnimationClip>,
-    names: &Query<&Name>,
-    transforms: &Query<&mut Transform>,
-    morphs: &Query<&mut MorphWeights>,
-    maybe_parent: Option<&Parent>,
-    parents: &Query<(Has<AnimationPlayer>, Option<&Parent>)>,
-    children: &Query<&Children>,
-) {
-    let paused = player.paused;
-    // Continue if paused unless the `AnimationPlayer` was changed
-    // This allow the animation to still be updated if the player.elapsed field was manually updated in pause
-    if paused && !player.is_changed() {
-        return;
-    }
-
-    // Apply the main animation
-    apply_animation(
-        1.0,
-        &mut player.animation,
-        paused,
-        root,
-        time,
-        animations,
-        names,
-        transforms,
-        morphs,
-        maybe_parent,
-        parents,
-        children,
-    );
-
-    // Apply any potential fade-out transitions from previous animations
-    for AnimationTransition {
-        current_weight,
-        animation,
-        ..
-    } in &mut player.transitions
-    {
-        apply_animation(
-            *current_weight,
-            animation,
-            paused,
-            root,
-            time,
-            animations,
-            names,
-            transforms,
-            morphs,
-            maybe_parent,
-            parents,
-            children,
-        );
+/// Removes animation transitions that have expired.
+///
+/// This needs to be done after applying the animations so that transitions
+/// don't miss applying their last frame on the frame they expire.
```

Comment:
This is only true if their weight is > 0. I made the comment on the other PR because in `bevy_tweening` there's no weight, so the animation always applies in full including on its last frame. But here with transitions and weights, do we really need that or can we remove any animation with 0 weight right away _before_ we apply them (since it won't contribute anyway, that's just a waste of CPU).

Reply:
Oh, good point. There's still an edge case where the main animation is paused and the last transition finishes, however: we have to make sure to re-apply the main animation.